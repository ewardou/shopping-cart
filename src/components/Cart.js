import React from 'react';
import cart from '../images/icons/cart.svg';
import '../style/Cart.css';

export default function Cart({ cartItems }) {
    function showCart(e) {
        console.log(e.target.childNodes);
        const [, , div] = e.target.childNodes;
        div.classList.toggle('visible');
    }

    function closeCart(e) {
        const div = e.target.parentNode;
        div.classList.toggle('visible');
    }

    function checkout() {
        alert('Thank you for your purchase :)');
    }

    const array = cartItems.map((item) => <p>{item}</p>);
    return (
        <span className="cart" onClick={showCart}>
            <span>0</span>
            <img src={cart} alt={'Shopping cart'} />
            <div>
                <button onClick={closeCart}>x</button>
                <div>
                    {cartItems.length !== 0 ? array : 'No items added yet'}
                </div>
                <button onClick={checkout}>Proceed to checkout</button>
            </div>
        </span>
    );
}
