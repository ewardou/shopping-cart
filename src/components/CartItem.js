import React from 'react';
import { info } from '../shoesInfo.js';
import '../style/CartItem.css';
import { useState, useEffect } from 'react';
// import forceUp

function Quantity({ removeItem, item, setTotalItems }) {
    const [quantity, setQuantity] = useState(1);
    function increase() {
        setQuantity((prev) => prev + 1);
        setTotalItems((prev) => prev + 1);
    }
    function decrease() {
        if (quantity <= 1) return removeItem();
        setQuantity((prev) => prev - 1);
        setTotalItems((prev) => prev - 1);
    }
    useEffect(() => {
        item.quantity = quantity;
    }, [quantity]);

    return (
        <div>
            <button onClick={decrease}>-</button>
            <span>{quantity}</span>
            <button onClick={increase}>+</button>
        </div>
    );
}

export function CartItem({ itemID, removeItem, setTotalItems }) {
    const [item] = info.filter((item) => item.itemID === itemID);
    return (
        <div className="card-item">
            <img src={item.image} alt={item.model} />
            <div>
                <p>{item.model}</p>
                <p>${item.price}</p>
                <Quantity
                    removeItem={() => removeItem(itemID)}
                    item={item}
                    setTotalItems={setTotalItems}
                />
            </div>
        </div>
    );
}
