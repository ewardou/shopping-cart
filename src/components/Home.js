import React from 'react';
import { Link } from 'react-router-dom';
import '../style/Home.css';
import fastIcon from '../images/icons/fast.svg';
import bestPriceIcon from '../images/icons/best-price.svg';
import newIcon from '../images/icons/new.svg';

function Quality({ title, url }) {
    return (
        <div className="quality">
            <img src={url} alt="" />
            <p>{title}</p>
            <p>Lorem ipsum dolor sit amet consectetur adipisicing elit.</p>
        </div>
    );
}

export function Home() {
    return (
        <div className="home">
            <div className="cover-background">
                <div className="cover-content">
                    <h1>Sneakers straight to your doorstep</h1>
                    <p>
                        Lorem ipsum dolor sit amet consectetur adipisicing elit.
                        Soluta odit possimus, aspernatur porro sint mollitia est
                        nostrum repudiandae esse officiis.
                    </p>
                    <Link to={'/shop'}>SHOP NOW</Link>
                </div>
            </div>
            <h1>You'll love shopping with us!</h1>
            <div className="qualities">
                <Quality title="Fast shipping" url={fastIcon} />
                <Quality title="Best prices" url={bestPriceIcon} />
                <Quality
                    title="Original models and new releases"
                    url={newIcon}
                />
            </div>
        </div>
    );
}
