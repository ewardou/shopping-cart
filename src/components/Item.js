import React from 'react';
import { useParams } from 'react-router-dom';

export default function Item() {
    const { itemID } = useParams();
    return <h1>{itemID} information</h1>;
}
