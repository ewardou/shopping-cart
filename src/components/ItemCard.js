import React from 'react';
import { Link } from 'react-router-dom';
import '../style/ItemCard.css';

export default function ItemCard(props) {
    function modifyImage(e) {
        const div = e.target.childNodes[0];
        div.childNodes[0].classList.toggle('hidden');
        div.childNodes[1].classList.toggle('hidden');
    }

    return (
        <div
            className="card"
            onMouseEnter={modifyImage}
            onMouseLeave={modifyImage}
        >
            <Link to={`${props.itemID}`}>
                <div>
                    <img src={props.image} alt={props.model} />
                    <img
                        src={props.secondImage}
                        alt={props.model}
                        className="secondImage hidden"
                    />
                </div>
                <p>{props.model}</p>
                <p>{`Price: $${props.price} `}</p>
            </Link>
        </div>
    );
}
