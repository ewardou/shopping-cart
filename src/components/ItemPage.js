import React from 'react';
import { useParams } from 'react-router-dom';
import '../style/ItemPage.css';
import { Link } from 'react-router-dom';
import { info } from '../shoesInfo';
import { useState } from 'react';

function Gallery({ item, onClick }) {
    return (
        <div className="gallery">
            <img src={item.image} alt={item.model} onClick={onClick} />
            <img src={item.secondImage} alt={item.model} onClick={onClick} />
            <img src={item.thirdImage} alt={item.model} onClick={onClick} />
            <img src={item.fourthImage} alt={item.model} onClick={onClick} />
        </div>
    );
}

export default function ItemPage() {
    const { itemID } = useParams();
    const [item] = info.filter((item) => {
        return item.itemID === itemID;
    });
    const [image, setImage] = useState(item.image);
    function changeImage(e) {
        setImage(e.target.src);
    }

    return (
        <div className="item-page">
            <Link to="/shop">Go back</Link>
            <div>
                <img src={image} alt={item.model} />
            </div>
            <div>
                <h1>{item.model}</h1>
                <p>{`$${item.price}`}</p>
                <Gallery item={item} onClick={changeImage} />
                <p>
                    <label htmlFor="size">Size: </label>
                    <select id="size">
                        <option>7</option>
                        <option>7.5</option>
                        <option>8</option>
                        <option>8.5</option>
                        <option>9</option>
                        <option>9.5</option>
                        <option>10</option>
                        <option>10.5</option>
                        <option>11</option>
                        <option>11.5</option>
                        <option>12</option>
                        <option>12.5</option>
                    </select>
                </p>
                <button>Add to cart</button>
            </div>
        </div>
    );
}
