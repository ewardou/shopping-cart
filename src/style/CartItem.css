.card-item {
    width: 100%;
    height: 110px;
    display: flex;
    font-size: 1rem;
    align-items: center;
}

.card-item > img {
    height: 100%;
}
