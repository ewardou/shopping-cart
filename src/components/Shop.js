import React from 'react';
import { Link } from 'react-router-dom';
import ItemCard from './ItemCard';
import fast from '../images/shoes/NB5740.webp';
import { info } from '../shoesInfo';

export function Shop() {
    return (
        <div>
            <h1>Hello from shop</h1>
            <ItemCard itemID={'item4'} model={'test model'} image={fast} />
        </div>
    );
}
