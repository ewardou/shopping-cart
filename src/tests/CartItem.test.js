import React from 'react';
import { render, screen } from '@testing-library/react';
import { CartItem } from '../components/CartItem';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

test('Renders component correctly', () => {
    jest.mock('../shoesInfo', () => {
        return [
            {
                itemID: 'item1',
                price: '50',
                quantity: '1',
                model: 'test model',
                image: '',
            },
        ];
    });

    const removeItem = jest.fn();
    const setTotalItems = jest.fn();
    const { container } = render(
        <CartItem
            itemID={'item1'}
            removeItem={removeItem}
            setTotalItems={setTotalItems}
        />
    );
    expect(container).toMatchSnapshot();
});

test("Pressing the '+' button increases the quantity", () => {
    jest.mock('../shoesInfo', () => {
        return [
            {
                itemID: 'item1',
                price: '50',
                quantity: '1',
                model: 'test model',
                image: '',
            },
        ];
    });

    const removeItem = jest.fn();
    const setTotalItems = jest.fn();
    render(
        <CartItem
            itemID={'item1'}
            removeItem={removeItem}
            setTotalItems={setTotalItems}
        />
    );
    const counter = screen.getByText('1');
    const plusButton = screen.getByText('+');

    userEvent.click(plusButton);
    userEvent.click(plusButton);
    expect(counter.innerHTML).toBe('3');
});

test('Functions passed in props are called', () => {
    jest.mock('../shoesInfo', () => {
        return [
            {
                itemID: 'item1',
                price: '50',
                quantity: '1',
                model: 'test model',
                image: '',
            },
        ];
    });

    const removeItem = jest.fn();
    const setTotalItems = jest.fn();
    render(
        <CartItem
            itemID={'item1'}
            removeItem={removeItem}
            setTotalItems={setTotalItems}
        />
    );
    const minusButton = screen.getByText('-');
    const plusButton = screen.getByText('+');

    userEvent.click(plusButton);
    userEvent.click(minusButton);
    userEvent.click(minusButton);

    expect(setTotalItems).toHaveBeenCalled();
    expect(removeItem).toHaveBeenCalled();
});
