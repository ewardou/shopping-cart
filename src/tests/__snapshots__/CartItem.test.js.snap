// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders component correctly 1`] = `
<div>
  <div
    class="card-item"
  >
    <img
      alt="Adidas D.O.N Issue 3"
      src="ADIDASDON.webp"
    />
    <div>
      <p>
        Adidas D.O.N Issue 3
      </p>
      <p>
        $
        75.99
      </p>
      <div
        class="quantity"
      >
        <span>
          Qty.
        </span>
        <button>
          -
        </button>
        <span>
          1
        </span>
        <button>
          +
        </button>
      </div>
    </div>
  </div>
</div>
`;
